Add a vmware-k8s-1.20 variant

zmrow opened the pull request:
**Issue number:**
Closes #1451 


**Description of changes:**
This change adds the necessary files for a VMware Kubernetes 1.20 variant. 

This variant is currently able to join Kubernetes clusters on VMware given the proper user data.


**Testing done:**
All nodes used the following user data:
```
[settings.kubernetes]
api-server = "<redacted>"
cluster-dns-ip = "<redacted>"
bootstrap-token = "<redacted>"
cluster-certificate = "abc123"

[settings.host-containers.admin]
enabled=true
user-data = "<redacted>"

[settings.host-containers.control]
enabled = true
user-data= "<redacted>"
```

Built a `vmware-k8s-1.20` OVA and uploaded to vSphere.  Booted 3 nodes using the above user-data via `guestinfo.userdata`.  The nodes boot and properly join a previously provisioned CAPV cluster.
* Ensure that `vmtoolsd` is running and green
* All taints are removed and hosts are schedulable.  All cluster pods running successfully on the nodes (calico, vsphere-csi, etc)
* Run a `busybox` pod
* Inspected journal logs; nothing alarming there
* Host containers work normally
  * SSH to the node's IP address to ensure that admin container works.  `sudo sheltie` works as normal
  * Use SSM to start a session on the control container.  Everything works as normal
* Sonobuoy testing passed.
```
Plugin: e2e
Status: passed
Total: 5667
Passed: 309
Failed: 0
Skipped: 5358

Plugin: systemd-logs
Status: passed
Total: 4
Passed: 4
Failed: 0
Skipped: 0
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
Add a vmware-k8s-1.20 variant

This change adds the necessary files for a VMware Kubernetes 1.20
variant.

zmrow commented:
^ Fix the Cargo.lock... oops :)

bcressey commented on packages/containerd/containerd-config-toml_vmware-k8s:
```diff
@@ -0,0 +1,31 @@
+version = 2
```
Can you try to make this file a symlink to `containerd-config-toml_aws-k8s`, and convert `containerd-config-toml_vmware-dev` to a symlink to `containerd-config-toml_aws-dev`?

It's OK if we list them separately in the spec and treat them as different files.

zmrow replied:
I tried this with `vmware-dev` previously and couldn't get the build to work.  Same here:

```
$ ls -lah packages/containerd/
...
lrwxrwxrwx.  1 mrowicki mrowicki   30 Apr 23 16:03 containerd-config-toml_vmware-k8s -> containerd-config-toml_aws-k8s


$ cargo make -e BUILDSYS_VARIANT=vmware-k8s-1.19
...
  #17 32.99 install: cannot stat '/home/builder/rpmbuild/SOURCES/containerd-config-toml_vmware-k8s': No such file or directory
...
```

tjkirch replied:
That's unfortunate, but don't give up!  Instead of symlinks, we can rename the files to not be so specific.  We only have two configs, one for k8s and one for everything else - dev and ECS variants.  What if we call them `containerd-config-toml_k8s` and `containerd-config-toml_basic` or something similar?

zmrow replied:
I think we should make this change in a separate PR as I'd rather not block this work on something that affects other variants. :)

tjkirch replied:
What if I do it as a separate PR merged before this one?

tjkirch replied:
https://github.com/bottlerocket-os/bottlerocket/pull/1538

tjkirch replied:
#1538 is merged, hopefully this is easily fixed now, just using the `containerd-config-toml_k8s` file :)

bcressey commented on variants/vmware-k8s-1.19/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+    "cni-plugins",
+    "kubelet-1.19",
+    "release",
+    "static-pods",
```
`static-pods` should not be listed, that's handled by the `os` package now.

zmrow replied:
Ah - good catch.

bcressey commented on variants/vmware-k8s-1.19/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+# Don't rebuild crate just because of changes to README.
+exclude = ["README.md"]
+
+[package.metadata.build-variant]
+image-format = "vmdk"
```
Needs console parameters added.

zmrow commented:
* Rebased on `develop` (sorry this rebase should have been in a different push to make the diff easier to read)
* Added `supported_arches` and `kernel-parameters`
* Removed `static-pods`

webern commented on packages/os/os.spec:
```diff
@@ -72,7 +73,9 @@ Requires: %{_cross_os}thar-be-updates
 Requires: %{_cross_os}updog
 
 %if %{_is_k8s_variant}
+%if %{_is_aws_variant}
 Requires: %{_cross_os}pluto
+%endif
 Requires: %{_cross_os}static-pods
 %endif
```
Would it be weird to use indentation for nested ifs in this language?

tjkirch replied:
Yep, not common.

webern replied:
:(

webern reviewed (commented):
🍰

tjkirch commented on variants/README.md:
```diff
@@ -69,6 +69,11 @@ The [vmware-dev](vmware-dev/Cargo.toml) variant has useful packages for local de
 It includes tools for troubleshooting as well as Docker for running containers.
 User data will be read from a mounted CD-ROM, either from a file named "user-data" or from an OVF file.
 
+### vmware-k8s-1.19: VMWare Kubernetes 1.19 node
```
Here and elsewhere: the company stylizes it as `VMware` with a lowercase "w".

tjkirch commented on variants/README.md:
```diff
@@ -69,6 +69,11 @@ The [vmware-dev](vmware-dev/Cargo.toml) variant has useful packages for local de
+
+The [vmware-k8s-1.19](vmware-k8s-1.19/Cargo.toml) variant includes the packages needed to run a Kubernetes worker node as a VMWare guest.
+It supports self-hosted clusters.
+
```
We should have the "This variant is compatible with" line, which I'm not putting as a suggestion yet because I assume we're updating to 1.20.

tjkirch commented on variants/vmware-k8s-1.19/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+supported-arches = ["x86_64"]
+kernel-parameters = [
+    "console=ttyS0",
+    "console=tty1",
```
Should probably add `systemd.unified_cgroup_hierarchy=1` like 1.20.  (I'm also assuming we'll update this variant to 1.20.)

tjkirch replied:
Or not!  https://github.com/bottlerocket-os/bottlerocket/pull/1533  :D

tjkirch commented on sources/models/src/vmware-k8s-1.19/defaults.d/50-vmware-k8s.toml:
```diff
@@ -0,0 +1,71 @@
+[configuration-files.containerd-config-toml]
```
This feels like a lot of duplication to me.  I think we could separate 50-aws-k8s.toml into configs for containerd, kubernetes-services, kubernetes-settings, and metrics, and then all of kubernetes-settings (the biggest piece) and metrics would be the same, with just the small kubernetes-settings and containerd configs changing...

zmrow replied:
I like the idea, but I'd rather not make those larger changes as part of this PR.

tjkirch replied:
What if I do it as a separate PR merged before this one?

tjkirch replied:
With https://github.com/bottlerocket-os/bottlerocket/pull/1542, instead of 50-vmware-k8s.toml, you should be able to link to the shared kubernetes-containerd.toml, kubernetes-services.toml, and public-ntp.toml files, and create a new shared kubernetes-vmware.toml file with the bits that remain.

tjkirch replied:
#1542 is merged, hopefully this would be easy now :)

tjkirch commented on sources/models/src/vmware-k8s-1.19/defaults.d/10-defaults.toml:
```diff
@@ -0,0 +1 @@
+../../../shared-defaults/defaults.toml
```
We should link to lockdown-integrity.toml too.

tjkirch commented on .github/workflows/build.yml:
```diff
@@ -20,6 +20,9 @@ jobs:
           - variant: vmware-dev
```
For lack of a better place to comment - we'll need the new variant in the variant list in the top-level README.md.

tjkirch commented on packages/os/os.spec:
```diff
@@ -1,5 +1,6 @@
 %global _cross_first_party 1
-%global _is_k8s_variant %(if echo %{_cross_variant} | grep -q "k8s"; then echo 1; else echo 0; fi)
+%global _is_k8s_variant %(if echo %{_cross_variant} | grep -Fq "k8s"; then echo 1; else echo 0; fi)
```
Why remove `-F`?  It's still a fixed string.

tjkirch replied:
I have no idea how I read this backwards, sorry!

zmrow commented:
Updated to use K8s 1.20 and integrated all the requested feedback.

#1510 is being worked by @bcressey separately and will have a separate PR to merge into this variant.

Marked this PR as ready for review.

tjkirch commented on sources/models/shared-defaults/vmware-host-containers.toml:
```diff
@@ -0,0 +1,14 @@
+# Both containers are disabled by default in VMWare because the user must
```
```suggestion
# Both containers are disabled by default in VMware because the user must
```

tjkirch commented on sources/models/shared-defaults/vmware-host-containers.toml:
```diff
@@ -0,0 +1,14 @@
+# Both containers are disabled by default in VMWare because the user must
+# supply user data in order to use the containers.  The admin container isn't
+# useful without SSH keys/CA certs, and the control container can only be used
+# with hybrid SSM off of AWS.  VMWare users might not want to use either of
```
```suggestion
# with hybrid SSM off of AWS.  VMware users might not want to use either of
```

tjkirch commented on variants/README.md:
```diff
@@ -76,6 +76,15 @@ The [vmware-dev](vmware-dev/Cargo.toml) variant has useful packages for local de
 
+### vmware-k8s-1.20: VMware Kubernetes 1.20 node
+
+The [vmware-k8s-1.20](vmware-k8s-1.20/Cargo.toml) variant includes the packages needed to run a Kubernetes worker node as a VMWare guest.
```
```suggestion
The [vmware-k8s-1.20](vmware-k8s-1.20/Cargo.toml) variant includes the packages needed to run a Kubernetes worker node as a VMware guest.
```

tjkirch commented on README.md:
```diff
@@ -63,6 +63,10 @@ We also have a variant designed to work with ECS, currently in preview:
 The `aws-k8s-1.15` variant is deprecated and will no longer be supported in Bottlerocket releases.
 We recommend users replace `aws-k8s-1.15` nodes with the [latest variant compatible with their cluster](variants/).
 
+Another variant we have in preview is designed to be a Kubernetes worker node in VMware:
```
nit: maybe put this above the deprecated list?  Best foot forward, etc.

tjkirch reviewed (commented):
I'm working on the config refactors for https://github.com/bottlerocket-os/bottlerocket/pull/1511#discussion_r622389780 and https://github.com/bottlerocket-os/bottlerocket/pull/1511#discussion_r622390048.

zmrow commented:
^ Addresses @tjkirch 's comments

zmrow commented:
^ rebase on `develop`, no changes.

zmrow commented:
^ Adds the kernel to `Cargo.toml` and updates `Cargo.lock`

tjkirch commented:
^ Wow, sorry.  I thought I was being funny by saying my PR would resolve a comment in this PR, but it closed the whole PR.  Reopened!

tjkirch commented on sources/models/src/vmware-k8s-1.20/defaults.d/50-vmware-k8s.toml:
```diff
@@ -0,0 +1,71 @@
+
+# Metrics
+[settings.metrics]
+service-checks = ["apiserver", "chronyd", "containerd", "host-containerd", "kubelet"]
```
A note so we don't forget to add vmtoolsd to this list after https://github.com/bottlerocket-os/bottlerocket/pull/1529

zmrow commented:
^ rebase on `develop`, no changes (yet)

zmrow commented:
^ Yet another rebase on develop, no changes (yet)

zmrow commented:
^ Addressed all the changes necessary after rebase and addressed remaining PR comments!
